### PR TITLE
SI-6240 Synchronizes Names 

### PR DESCRIPTION
**Pull request by retronym**
Previously we didn't have all possible name creation facilities covered
with locks, so some of them silently misbehaved and caused much grief:
http://groups.google.com/group/scala-internals/browse_thread/thread/ec1d3e2c4bcb000a.

This patch gets all the name factories under control.

Rather than relying on subclasses to override mutating methods
and add synchronization, they should instead override `synchronizeNames`
to return true and leave the placement of the locks up to
`reflect.internal.Names`.

These locks are placed around sections that mutate `typeHashtable`
and `termHashtable`.

This is done in the reflection universe, and in the interactive
compiler. The latter change will obviate an incomplete attempt
do to the same in `ScalaPresentationCompiler` in the scala-ide
project.

Review by @xeno-by @gkossakowski

### CODE REVIEW

**Comment by retronym** [comment]
Note: don't close SI-6240 if this is merged, this is just a small part of the puzzle, but a part that will help us to reduce one source of flakiness in IDE tests.

**Comment by retronym** [comment]
Ping @xeno-by @gkossakowski

File: src/reflect/scala/reflect/internal/Names.scala [hunk 1]
```diff
@@ -18,6 +18,18 @@ trait Names extends api.Names {
...
+  // kick in in the compiler universe, where access to the lock is single threaded. But,
+  // objects created in the first 4seconds of the JVM startup aren't eligible for biased
+  // locking.
+  //
+  // We might also be able to accept the performance hit, but we don't have tools to
+  // detect performance regressions.
+  //
+  // Discussion: https://groups.google.com/forum/#!search/biased$20scala-internals/scala-internals/0cYB7SkJ-nM/47MLhsgw8jwJ
+  protected def synchronizeNames: Boolean = false
+  private val nameLock: Object = new {}
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Names.scala]: How about `new Object`?

File: src/reflect/scala/reflect/internal/Names.scala [hunk 2]
```diff
@@ -64,61 +76,68 @@ trait Names extends api.Names {
...
     }
+    if (synchronizeNames) nameLock.synchronized(body) else body
   }
-  protected def newTypeName(cs: Array[Char], offset: Int, len: Int, cachedString: String): TypeName =
+
+  final def newTypeName(cs: Array[Char], offset: Int, len: Int, cachedString: String): TypeName =
     newTermName(cs, offset, len, cachedString).toTypeName
 
   /** Create a term name from string. */
+  @deprecatedOverriding("To synchronize, use `override def synchronizeNames = true`", "2.11.0") // // overriden in https://github.com/scala-ide/scala-ide/blob/master/org.scala-ide.sdt.core/src/scala/tools/eclipse/ScalaPresentationCompiler.scala
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Names.scala]: Extraneous `//`

Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/Names.scala]: +1

File: src/reflect/scala/reflect/internal/Names.scala (hunk 1 above)
Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/Names.scala]: I find `new Object` or `new AnyRef` more readable as well.

**Comment by gkossakowski** [comment]
@retronym: Apart from minor nitpicks this looks good to me.

**Comment by xeno-by** [comment]
Same here. LGTM

**Comment by retronym** [comment]
LGTM echo (for the bot).

The minor comments are addressed in latest commit.

**Comment by retronym** [comment]
@xeno-by Regarding the rest of SI-6240, I think we can avoid the lazy vals for locks by defining the lock object in the superclasses beside the thing it is protecting, like `nameLock` here. If someone tried to use `nameLock` before it is initialized, it is of no concern, because before they would just have used the `chrs` field before it was initialized, so we are no worse off.

Even if, unlike here, the locking itself is done in subclasses in some other places, we are still free to define the lock in the superclass if it makes reasoning about initialization order easier.

**Comment by dragos** [comment]
Would it be possible to backport this to 2.10.x?

**Comment by retronym** [comment]
@dragos Conservatively, I'd like to see how it plays out in master for a while (the same for the remaining parts of reflection sync.) But we might be able to get all of that backported for 2.10.4.

**Comment by dragos** [comment]
@retronym What are we looking for before backporting?

**Comment by retronym** [comment]
As discussed, let's wait a few days to see if all signs of name-table corruption have dried up from our IDE builds against master. Perhaps you could schedule a run of extra builds to get more confidence. Then we'll backport around the end of the week.

**Comment by retronym** [comment]
Backport: https://github.com/retronym/scala/compare/scala:2.10.x...backport/6240-name-sync?expand=1

Building: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1001/console

**Pull request merged**